### PR TITLE
MOTS-204: Add endpoint to handle user password change

### PR DESCRIPTION
**Pull request by kcissewski**

### CODE REVIEW

File: src/main/java/org/motechproject/mots/service/UserService.java [hunk 1]
```diff
@@ -38,11 +39,29 @@ public User getUser(UUID id) {
         new EntityNotFoundException("User with id: {0} not found", id.toString()));
   }
 
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_ROLE)
+  public User getUserByUserName(String userName) {
+    Optional<User> optionalUser = userRepository.findOneByUsername(userName);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/UserService.java]: you should use orElseThrow and throw EntityNotFoundException here

File: src/main/java/org/motechproject/mots/service/UserService.java [hunk 2]
```diff
@@ -38,11 +39,29 @@ public User getUser(UUID id) {
         new EntityNotFoundException("User with id: {0} not found", id.toString()));
   }
 
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_ROLE)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/UserService.java]: if you use Manage users role here, only admin will be able to change his password, remove this

File: src/main/java/org/motechproject/mots/service/UserService.java [hunk 3]
```diff
@@ -38,11 +39,29 @@ public User getUser(UUID id) {
...
   public User saveUser(User user) {
     return userRepository.save(user);
   }
 
+  /**
+   * Updates user's password.
+   * @param user which password is about to change.
+   * @param newPassword is new password value for user.
+   */
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_ROLE)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/UserService.java]: remove the authorization

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 4]
```diff
@@ -86,6 +89,35 @@ public UserDto saveUser(@PathVariable("id") UUID id, @RequestBody UserDto userDt
...
+      method = RequestMethod.POST)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public ResponseEntity<String> changeCurrentUserPassword(
+      @PathVariable("currentPassword") String currentPassword,
+      @PathVariable("newPassword") String newPassword) {
+
+    String userName =
+        (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+    User existingUser = userService.getUserByUserName(userName);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/UserController.java]: move all of this to service, change the user parameter in changeUserPassword to user name and add old passowrd parameter

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 5]
```diff
@@ -86,6 +89,35 @@ public UserDto saveUser(@PathVariable("id") UUID id, @RequestBody UserDto userDt
...
+  @RequestMapping(value = "/user/password/change/{currentPassword}/{newPassword}",
+      method = RequestMethod.POST)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public ResponseEntity<String> changeCurrentUserPassword(
+      @PathVariable("currentPassword") String currentPassword,
+      @PathVariable("newPassword") String newPassword) {
+
+    String userName =
+        (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/UserController.java]: @jkondrat do you think this is the best way to get the logged in user?

File: src/main/java/org/motechproject/mots/validate/validators/PasswordValidator.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.motechproject.mots.validate.validators;
+
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+
+public class PasswordValidator {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/validate/validators/PasswordValidator.java]: I think you can move this logic to user service and remove this validator

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 7]
```diff
@@ -86,6 +89,35 @@ public UserDto saveUser(@PathVariable("id") UUID id, @RequestBody UserDto userDt
...
+
+    if (existingUser == null
+        || !PasswordValidator.validateCurrentPassword(
+            existingUser.getPassword(), currentPassword)) {
+      return new ResponseEntity<>("Current password is incorrect.", HttpStatus.BAD_REQUEST);
+    }
+
+    userService.changeUserPassword(existingUser, newPassword);
+
+    return new ResponseEntity<>("Password changed successfully.", HttpStatus.OK);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/UserController.java]: we're not using the response entity in MOTS, return a String, you already set a status with annotation

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 8]
```diff
@@ -86,6 +89,35 @@ public UserDto saveUser(@PathVariable("id") UUID id, @RequestBody UserDto userDt
...
+      @PathVariable("newPassword") String newPassword) {
+
+    String userName =
+        (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+    User existingUser = userService.getUserByUserName(userName);
+
+    if (existingUser == null
+        || !PasswordValidator.validateCurrentPassword(
+            existingUser.getPassword(), currentPassword)) {
+      return new ResponseEntity<>("Current password is incorrect.", HttpStatus.BAD_REQUEST);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/UserController.java]: when you move this to the service you can throw IllegalArgmentException with this message and maybe change "Current" to "Old"

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 9]
```diff
@@ -86,6 +89,35 @@ public UserDto saveUser(@PathVariable("id") UUID id, @RequestBody UserDto userDt
     return userMapper.toDto(userService.saveUser(existingUser));
   }
 
+  /**
+   * Update User password.
+   * @param currentPassword user previous password, should be the same as current password
+   * @param newPassword user's new password
+   * @return updated User
+   */
+  @RequestMapping(value = "/user/password/change/{currentPassword}/{newPassword}",
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/UserController.java]: You should send those passwords as a parameters not in the url

File: src/main/java/org/motechproject/mots/web/UserController.java (hunk 5 above)
Comment (jkondrat) [reply src/main/java/org/motechproject/mots/web/UserController.java]: It's fine to leave it as is, but we could also add principal to the parameters like so:
> changeCurrentUserPassword(Principal principal) {
>   principal.getName()
> }

File: src/main/java/org/motechproject/mots/service/UserService.java [hunk 10]
```diff
@@ -43,6 +43,24 @@ public User saveUser(User user) {
...
+  /**
+   * Updates user's password.
+   * @param username of user which password is about to change.
+   * @param newPassword is new password value for user.
+   * @param oldPassword is current user's password.
+   */
+  public void changeUserPassword(String username, String newPassword, String oldPassword) {
+    User user = getUserByUserName(username);
+
+    if (user == null || !passwordsMatch(oldPassword, user.getPassword())) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/UserService.java]: User will never be null, exception will be thrown when user will not be found

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 11]
```diff
@@ -86,6 +88,25 @@ public UserDto saveUser(@PathVariable("id") UUID id, @RequestBody UserDto userDt
     return userMapper.toDto(userService.saveUser(existingUser));
   }
 
+  /**
+   * Update User password.
+   * @param oldPassword user previous password, should be the same as current password
+   * @param newPassword user's new password
+   */
+  @RequestMapping(value = "/user/password/change",
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/UserController.java]: I think this should fit in one line

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 12]
```diff
@@ -86,6 +88,25 @@ public UserDto saveUser(@PathVariable("id") UUID id, @RequestBody UserDto userDt
...
+      method = RequestMethod.POST)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public String changeCurrentUserPassword(@RequestParam("oldPassword") String oldPassword,
+      @RequestParam("newPassword") String newPassword) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/UserController.java]: Did you try to add Principal as a parameter here?
public String changeCurrentUserPassword(@RequestParam("oldPassword") String oldPassword, @RequestParam("newPassword") String newPassword, Principal principal)

**Pull request merged**